src/server/rewindable.rs

```diff
@@ -0,0 +1,204 @@
+use bytes::{Buf, BufMut, Bytes, IntoBuf};
+use futures::{Async, Poll};
+use std::io::{self, Read, Write};
+use tokio_io::{AsyncRead, AsyncWrite};
+
+#[derive(Debug)]
+pub struct Rewindable<T> {
+    pre: Option<Bytes>,
+    inner: T,
+}
+
+impl<T> Rewindable<T> {
+    pub(super) fn new(tcp: T) -> Rewindable<T> {
+        Rewindable {
+            pre: None,
+            inner: tcp,
+        }
+    }
+    pub fn prepend(&mut self, bs: Bytes) {
+        debug_assert!(self.pre.is_none());
+        self.pre = Some(bs);
+    }
+    pub fn into_inner(self) -> T {
+        self.inner
+    }
+}
+
+impl<T> Read for Rewindable<T>
+where
+    T: Read,
+{
+    #[inline]
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        if let Some(pre_bs) = self.pre.take() {
+            // If there are no remaining bytes, let the bytes get dropped.
+            if pre_bs.len() > 0 {
+                let mut pre_reader = pre_bs.into_buf().reader();
+                let mut read_cnt = pre_reader.read(buf)?;
+
+                let mut new_pre = pre_reader.into_inner().into_inner();
+                new_pre.advance(read_cnt);
+
+                // Put back whats left
+                if new_pre.len() > 0 {
+                    self.pre = Some(new_pre);
+                }
+
+                if read_cnt < buf.len() {
+                    let (_, tail) = buf.split_at_mut(read_cnt);
+                    read_cnt += self.inner.read(tail)?;
+                }
+                return Ok(read_cnt);
+            }
+        }
+        self.inner.read(buf)
+    }
+}
+
+impl<T> Write for Rewindable<T>
+where
+    T: Write,
+{
+    #[inline]
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        self.inner.write(buf)
+    }
+
+    #[inline]
+    fn flush(&mut self) -> io::Result<()> {
+        self.inner.flush()
+    }
+}
+
+impl<T> AsyncRead for Rewindable<T>
+where
+    T: AsyncRead,
+{
+    #[inline]
+    unsafe fn prepare_uninitialized_buffer(&self, buf: &mut [u8]) -> bool {
+        self.inner.prepare_uninitialized_buffer(buf)
+    }
+
+    #[inline]
+    fn read_buf<B: BufMut>(&mut self, buf: &mut B) -> Poll<usize, io::Error> {
+        if let Some(bs) = self.pre.take() {
+            // If there are no remaining bytes, let the bytes get dropped.
+            if bs.len() > 0 {
+                let mut reader = bs.into_buf().reader();
```

Comment:
I think there may be a more idiomatic way to copy between a `Buf` and a `BufMut`. I think this may work:

```rust
let cnt = buf.remaining_mut();
let mut bs = bs.into_buf().take(cnt);
buf.put(&mut bs);

// bs.into_inner() stuff
```

It's unfortunate the dance needed for `BytesMut`. It's already been decided that it was mistake that it doesn't implement `Buf` itself, but it can't without a breaking change. It will in bytes 0.5.

Reply:
Excellent info as always.